File: bongf/python/ch10_graph/ch10_9_curriculum.py
from collections import deque
import copy 

def solution() :
    n = int(input())
    indegree = [0] * (n+1)
    hour = [0] * (n+1)
    graph = [[] for _ in range(n+1)]

    for i in range(1, n+1) :
        info = list(map(int, input().split()))
        hour[i] = info[0]

        # dobin : 
        # for i in info[1:-1] : 
        #   indegree[i] += 1
        #   graph[i].append(i)
        for x in range(1, len(info)-1):
            graph[info[x]].append(i)
            indegree[i] += 1

    result = copy.deepcopy(hour)
    q = deque()
    for i in range(1, n+1):
        if indegree[i] == 0 :
            q.append(i)

    while q :
        now = q.popleft()
        for i in graph[now]:
            result[i] = max(result[i], result[now] + hour[i])
            indegree[i] -= 1
            if indegree[i] == 0 :
                q.append(i)

    for i in range(1, n+1):
        print(result[i])

solution()

'''
5
10 -1
10 1 -1
4 1 -1
4 3 1 -1
3 3 -1
결과 
10
20
14
18
17
'''

File: bongf/java/src/dongbinbook/ch10_graph/Ch10_9_Curriculum.java
package dongbinbook.ch10_graph;

import java.util.*;

public class Ch10_9_Curriculum {
    public static void main(String[] args) {
        Ch10_9_Curriculum c = new Ch10_9_Curriculum();
        System.out.println(Arrays.toString(c.solution()));
    }

    public int[] solution() {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        sc.nextLine();
        int[] time = new int[n+1];
        int[] indegree = new int[n+1];

       List<List<Integer>> graph = new ArrayList<>();
       graph.add(new ArrayList<>());

        for (int i = 1; i < n+1; i++) {
            Integer[] in = Arrays.stream(sc.nextLine().split(" "))
                    .map(Integer::valueOf)
                    .toArray(Integer[]::new);

            time[i] = in[0];
            graph.add(new ArrayList<>());
            for (int j = 1; j < in.length-1; j++) {
                graph.get(in[j]).add(i);
                indegree[i]++;
            }
        }

        Queue<Integer> q = new LinkedList<>();
        for (int j = 1; j <= n; j++) {
            if(indegree[j] == 0) {
                q.offer(j);
            }
        }

        int[] result = Arrays.copyOf(time, time.length);
        while(!q.isEmpty()) {
            int now = q.poll();
            for(  int x : graph.get(now)) {
                result[x] = Math.max(result[x], time[x] + result[now]);
                indegree[x]--;
                if(indegree[x] == 0) {
                    q.offer(x);
                }
            }
        }
        return Arrays.copyOfRange(result, 1, result.length);
    }
}

/*
5
10 -1
10 1 -1
4 1 -1
4 3 1 -1
3 3 -1
결과
10
20
14
18
17
 */
